Userland/Libraries/LibCrypto/Checksum/CRC16.h

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (c) 2023, kleines Filmröllchen <filmroellchen@serenityos.org>.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Format.h>
+#include <AK/Types.h>
+#include <LibCrypto/Checksum/ChecksumFunction.h>
+
+namespace Crypto::Checksum {
+
+// A generic 16-bit Cyclic Redundancy Check.
+// Note that as opposed to CRC32, this class operates big-endian with MSB first, so the polynomial must not be reversed.
+// For example, the polynomial x¹⁶ + x¹² + x⁵ + 1 is represented as 0x1021 and not 0x8408.
```

Comment:
This seems a little concerning for usability, could we pick a user-facing endianness and byteswap it at compile time if needed, making the polynomial endianness an internal implementation detail?

It would really just look like `static constexpr auto actual_polynomial = AK::HostIsBigEndian ? polynomial : __builtin_bswap16(polynomial);` as a private class member.

Reply:
The thing is you need to *bitwise reverse* the polynomial between LE and BE. Our CRC32 impl also assumes LE, even though it's not generic at the moment. So I don't see the benefit; my comment should be enough so the user copies the BE poly and not the LE.

Reply:
A comment on top of a template will not help anyone.
I'm not saying you should switch to LE, in saying the two implementations should present the same endianness to the user. Consider it a nitpick.

Reply:
Yes, but the problem remains, I need to *bitswap*, not byteswap.

Reply:
Sure, but the CRC8 also operates in MSB/big endian form, and it was merged without complaints. None of the two endiannesses, even on the bit level, is superior, I’m using big-endian here since that’s what you get when you port the byte-level algorithm to 16 bits directly, and we’re using little-endian in CRC32 since we have hardware implementations and the little-endian slicing-by-8.

Reply:
There’s a reason this representation of polynomials is called “Reversed”, it’s not the first representation you can find online, and it’s not the one considered standard.

Reply:
I have no qualms with the implementation using one or the other endianness, I just want the APIs to be uniform.

If this form really is the canonical one for this specific CRC, and all/most of the CRC16s you'll find in specs are in this form, fine, feed free to ignore the suggestion - but please at least call the template parameter something like `big_endian_polynomial` if you do that.